File: benchmarks/go/12_bigint_overflow.go
// BigInt Overflow Path Benchmark - Go Implementation

package main

import (
	"fmt"
	"math/big"
)

func main() {
	value := new(big.Int).Lsh(big.NewInt(1), 100)
	addend := big.NewInt(123456789)
	subtrahend := big.NewInt(98765432)
	modulus := big.NewInt(97)
	checksum := int64(0)
	tmp := new(big.Int)

	for i := 0; i < 200000; i++ {
		value.Add(value, addend)
		value.Sub(value, subtrahend)
		tmp.Mod(value, modulus)
		checksum += tmp.Int64()
	}

	fmt.Printf("bigint overflow checksum: %d\n", checksum)
}

File: benchmarks/rust/12_bigint_overflow.rs
// BigInt Overflow Path Benchmark - Rust Implementation

fn main() {
    let mut value: i128 = 1i128 << 100;
    let mut checksum: i64 = 0;

    for _ in 0..200_000 {
        value += 123_456_789;
        value -= 98_765_432;
        checksum += (value % 97) as i64;
    }

    println!("bigint overflow checksum: {}", checksum);
}
